File: packages/typescript/src/emitter/comments.ts
import { YAMLMap, YAMLSeq, Scalar, Pair } from "yaml";
import { renderBlockComment, renderEolComment } from "./comment-geometry.js";

/**
 * Comment attachment: the single owner of how a comment lands on a `yaml` node.
 *
 * Every path by which a YAML comment reaches a node routes through this module:
 *
 * - {@link attachFieldComment} — a block/EOL comment from a `Commented` wrapper
 *   on one mapping field, attached to that field's pair.
 * - {@link attachModelComment} — a model's OWN block/EOL comment, rendered on
 *   the map as a whole. `atSeqItem` is the one container decision that the old
 *   duplicate-comment workaround hacked around.
 *
 * This module decides WHICH node a comment lands on; `emitter/comment-geometry.ts`
 * decides what the payload looks like. Every write below goes through a
 * renderer from that module — nothing here writes raw comment text.
 *
 * This is the TypeScript peer of `emitter/comments.py`.
 */

/** The pair's key as a `Scalar`, promoting a non-`Scalar` key in place. */
function keyScalar(pair: Pair): Scalar {
  const key = pair.key instanceof Scalar ? pair.key : new Scalar(pair.key);
  pair.key = key;
  return key;
}

/** Add a block comment before *pair*'s key, after any block comment already there. */
function addBlockComment(pair: Pair, text: string): void {
  const key = keyScalar(pair);
  const rendered = renderBlockComment(text);
  key.commentBefore = key.commentBefore ? `${key.commentBefore}\n${rendered}` : rendered;
}

/**
 * Put an EOL comment on *pair*, or — when the payload cannot sit at
 * end-of-line — above it as a block comment.
 *
 * The EOL comment goes on the value for scalars and on the KEY for complex
 * values (map / seq), so it renders on the key line rather than being
 * swallowed into the nested block. That container rule is the same in both
 * containers and in both ports; it is also what keeps `lineComment`'s
 * no-column branch out of the reachable set (see `comment-geometry.ts`).
 */
function addEolComment(pair: Pair, text: string): void {
  const rendered = renderEolComment(text);
  if (rendered === null) {
    // Multi-line payload: it cannot sit at end-of-line in either backend, so
    // it degrades to a block comment on the same item.
    addBlockComment(pair, text);
    return;
  }
  if (pair.value instanceof YAMLMap || pair.value instanceof YAMLSeq) {
    keyScalar(pair).comment = rendered;
  } else if (pair.value instanceof Scalar) {
    pair.value.comment = rendered;
  } else {
    const val = new Scalar(pair.value);
    val.comment = rendered;
    pair.value = val;
  }
}

/**
 * Attach a block and/or EOL comment from a `Commented` wrapper to one mapping
 * field's pair.
 *
 * - block comment → `key.commentBefore` (renders above the field).
 * - EOL comment → on the value for scalars; on the key for complex values
 *   (map / seq) so it renders on the key line rather than swallowed by the
 *   nested block; above the field as a block comment when the payload is
 *   multi-line.
 */
export function attachFieldComment(pair: Pair, comment?: string, eolComment?: string): void {
  if (comment) {
    addBlockComment(pair, comment);
  }
  if (eolComment) {
    addEolComment(pair, eolComment);
  }
}

/**
 * Attach a model's OWN block/EOL comment, rendered on the map as a whole.
 *
 * `atSeqItem` is the one decision the old duplicate-comment workaround hacked
 * around — a model renders differently depending on its container:
 *
 * - `false` → the map is a field value or the document root: the block comment
 *   goes on the first key (`key.commentBefore`) and the EOL comment on the last
 *   entry's line.
 * - `true` → the map is a `- ` list entry: the block comment goes on the map
 *   node itself (`map.commentBefore`) so it renders above the dash, and the EOL
 *   comment goes on the FIRST entry's line so it renders on the dash/entry-point
 *   line (the peer of ruamel's seq-index EOL redirect to the item's first key).
 *
 * In both containers the EOL comment lands on the target pair's KEY when that
 * pair's value is a map or a seq — the same rule {@link attachFieldComment}
 * applies — so a `job({ eolComment })`, whose last key is `steps`, renders
 * `steps:  # …` rather than orphaning the comment below the step list.
 *
 * The container decision is made ONCE, by the caller — so no second call site
 * has to reverse a wrongly-placed attachment.
 */
export function attachModelComment(
  map: YAMLMap,
  comment: string | undefined,
  eolComment: string | undefined,
  opts: { atSeqItem: boolean },
): void {
  if (comment) {
    if (opts.atSeqItem) {
      map.commentBefore = renderBlockComment(comment);
    } else {
      const firstPair = map.items[0] as Pair | undefined;
      if (firstPair) {
        // A model's own comment precedes any field comment already on that key.
        const key = keyScalar(firstPair);
        const rendered = renderBlockComment(comment);
        key.commentBefore = key.commentBefore ? `${rendered}\n${key.commentBefore}` : rendered;
      }
    }
  }

  if (eolComment) {
    // Seq item → the entry-point (first) line; map value / root → the last
    // entry's line. This mirrors ruamel's seq-index EOL redirect on the Python
    // side.
    const target = (opts.atSeqItem ? map.items[0] : map.items[map.items.length - 1]) as
      | Pair
      | undefined;
    if (target) {
      addEolComment(target, eolComment);
    }
  }
}

File: packages/python/src/ghagen/emitter/comments.py
"""Comment attachment: the single owner of how a comment lands on a ruamel node.

Every path by which a YAML comment reaches a node routes through this module:

- :func:`attach` — the low-level primitive. Attaches a block and/or EOL comment
  to a map key or a seq index. Owns the map-vs-seq dispatch and the seq-item
  map EOL-redirect quirk. Comment *columns* are not its business:
  :mod:`ghagen.emitter.comment_geometry` owns those.
- :func:`attach_model_comment` — a model's OWN comment, rendered on the map as a
  whole: block before the first key, EOL after the last value. Used for the
  Document root AND every nested map-value model.

This is the Python peer of ``emitter/comments.ts``.
"""

from __future__ import annotations

from ruamel.yaml.comments import CommentedMap, CommentedSeq

from ghagen.emitter.comment_geometry import render_eol_comment


def _fits_at_eol(text: str) -> bool:
    """True when *text* can be rendered as an end-of-line comment.

    Delegates to :func:`~ghagen.emitter.comment_geometry.render_eol_comment`,
    which owns the rule; a multi-line payload cannot sit at end of line (ruamel
    would emit unparseable YAML) and is redirected to the block path.
    """
    return render_eol_comment(text) is not None


def attach(
    parent: CommentedMap | CommentedSeq,
    key: str | int,
    *,
    comment: str | None = None,
    eol_comment: str | None = None,
) -> None:
    """Attach comments to a key in a CommentedMap or index in a CommentedSeq.

    Args:
        parent: The CommentedMap or CommentedSeq to attach comments to.
        key: The key (for maps) or index (for sequences) to attach comments to.
        comment: Block comment to place before the key.
        eol_comment: End-of-line comment to place after the value.

    Notes:
        When the target is a sequence index whose item is a non-empty
        ``CommentedMap``, ``eol_comment`` is redirected to the inner map's
        first key so the comment renders inline with that key rather than on
        the dash line (a ruamel.yaml quirk when EOL-commenting seq indices
        whose items are maps).

        A multi-line ``eol_comment`` cannot sit at end of line at all — ruamel
        would emit unparseable YAML — so it degrades to a block comment on the
        same item, joined after any ``comment`` already given. The rule lives
        in :mod:`ghagen.emitter.comment_geometry`.

        This module decides WHICH node a comment lands on; every column
        decision — the block-comment column and the end-of-line gutter — is
        made by :func:`~ghagen.emitter.comment_geometry.apply_comment_geometry`
        during :func:`~ghagen.emitter.yaml_writer.dump_yaml`.
    """
    # A payload that cannot sit at end of line degrades to a block comment on
    # the same item, after any block comment already there.
    if eol_comment is not None and not _fits_at_eol(eol_comment):
        comment = eol_comment if comment is None else f"{comment}\n{eol_comment}"
        eol_comment = None

    if isinstance(parent, CommentedMap) and isinstance(key, str):
        if comment is not None:
            parent.yaml_set_comment_before_after_key(key, before=comment)
        if eol_comment is not None:
            parent.yaml_add_eol_comment(eol_comment, key=key)
        return

    if isinstance(parent, CommentedSeq) and isinstance(key, int):
        if comment is not None:
            parent.yaml_set_comment_before_after_key(key, before=comment, indent=0)

        if eol_comment is not None:
            try:
                item = parent[key]
            except IndexError:
                item = None
            if isinstance(item, CommentedMap) and len(item) > 0:
                first_key = next(iter(item.keys()))
                item.yaml_add_eol_comment(eol_comment, key=first_key)
            else:
                parent.yaml_add_eol_comment(eol_comment, key=key)


def attach_model_comment(
    node: CommentedMap,
    *,
    comment: str | None = None,
    eol_comment: str | None = None,
) -> None:
    """Attach a model's OWN comment, rendered on the map as a whole.

    A block ``comment`` is placed before the map's first key; an
    ``eol_comment`` is placed after the map's last value (its last key's line).
    Used for the Document root AND every nested map-value model, so a model's
    own comment renders regardless of the container it sits in.

    A no-op on an empty map (no key to anchor the comment to).
    """
    if not node:
        return
    keys = list(node.keys())
    if not keys:
        return
    if comment is not None:
        attach(node, keys[0], comment=comment)
    if eol_comment is not None:
        attach(node, keys[-1], eol_comment=eol_comment)
